### PR TITLE
Add new direct Redis transport that doesn't use ASGI

### PR DESCRIPTION
**Pull request by nickwilliams-eventbrite**
Other improvements:
- Adds serialization in transport (Redis only for now)
- Recurses the entire config tree looking for `BasicClassSchema` configs that need resolving
- Adds type validation to ensure `BasicClassSchema` uses are proper
- Adds code coverage for better understanding where our weaknesses are
- Improves code coverage of settings
- Fixes lots of PEP8 errors

Still to do (in a second commit after this):
- Add a generic metrics recorder

Still to do (after rolling out all uses to the new Redis transport):
- Remove ASGI transport
- Remove serialization in higher layer (server/client)

### CODE REVIEW

**Comment by bgreenberg-eb** [comment]
Thanks for doing this! Looks good to me on a high level; I'll look at it in more detail tonight.

File: pysoa/common/transport/redis_gateway/backend/base.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        self._call(keys=[queue_key], args=[expiry, capacity, message], connection=connection)
+
+
+class BaseRedisClient(object):
+    DEFAULT_RECEIVE_TIMEOUT = 5
+    RESPONSE_QUEUE_SPECIFIER = '!'
+
+    def __init__(self, ring_size):
+        self._ring_size = ring_size
+        self._receive_index_generator = itertools.cycle(range(self._ring_size))  # may be overridden by subclasses
```
Comment (andrewgodwin) [inline pysoa/common/transport/redis_gateway/backend/base.py]: `_receive_index_generator` is never used.

File: pysoa/common/transport/redis_gateway/backend/sentinel.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        self._master_connections = {}
+
+        self._sentinel = redis.sentinel.Sentinel(self._setup_hosts(hosts), **(connection_kwargs or {}))
+        if sentinel_services:
+            self._validate_service_names(sentinel_services)
+            self._services = sentinel_services
+        else:
+            self._services = self._get_service_names()
+        self._ring_size = len(self._services)
+        self._receive_index_generator = itertools.cycle(range(self._ring_size))
```
Comment (andrewgodwin) [inline pysoa/common/transport/redis_gateway/backend/sentinel.py]: `_receive_index_generator` is never used.

File: pysoa/common/transport/redis_gateway/backend/base.py (hunk 1 above)
Comment (nickwilliams-eventbrite) [reply pysoa/common/transport/redis_gateway/backend/base.py]: Good catch. Fixed.

File: pysoa/common/transport/redis_gateway/backend/sentinel.py (hunk 2 above)
Comment (nickwilliams-eventbrite) [reply pysoa/common/transport/redis_gateway/backend/sentinel.py]: Good catch. Fixed.

File: pysoa/common/settings.py [hunk 3]
```diff
@@ -72,18 +91,18 @@ class BaseSettings(Settings):
...
     provides a measure of convenience while discouraging deep inheritance structures.
 
     To use Settings, instantiate the class with the raw settings value, and then
-    access the items using dict syntax - e.g. settings_instance["transport"]. The class
+    access the items using dict syntax - e.g. settings_instance['transport']. The class
     will merge any passed values into its defaults.
 
     You can override how certain fields are set by defining a method called
-    `convert_fieldname`.
+    `convert_[field_name]`.
```
[the remark this answers is not in the document]
Comment (nickwilliams-eventbrite) [reply pysoa/common/settings.py]: Using replacement bracket syntax `[]` to make this more descriptive, and fixed "fieldname" typo that my IDE was bugging me about.

[the remark this answers is not in the document]
Comment (vartec) [reply pysoa/common/settings.py]: Hmm.. it this case, maybe `convert_{field_name}` would be less confusing.

File: pysoa/common/transport/redis_gateway/backend/base.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        Returns the correct connection for the current thread. Pass `index` to use a server based on consistent hashing
+        of the key value; `pass` None to use a random server instead.
+
+        :param index: The optional index of a server to use
+        :return: the connection to use.
+        """
+        raise NotImplementedError()
+
+    def _get_random_index(self):
```
[the remark this answers is not in the document]
Comment (nickwilliams-eventbrite) [reply pysoa/common/transport/redis_gateway/backend/base.py]: No, it's also used right above in `get_connection`. The `get_connection` method handles picking the right connection index, which doesn't care about standard vs. Sentinel Redis. It then calls `_get_connection`, which has different implementation for standard vs. Sentinel Redis.

File: pysoa/common/settings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickwilliams-eventbrite) [reply pysoa/common/settings.py]: Ok. Done.

**Comment by nickwilliams-eventbrite** [comment]
>Really not a fan of this mock-based testing considering the amount of boilerplate it adds, and that it needs an external system tool installed (lua) anyway. But, I guess if you'll be the one maintaining it, I don't care enough to block.

I feel like requiring a functioning Redis server (which itself requires Lua to be installed) for the purposes of unit testing is even more burdensome than requiring the Lua library to be installed. I'm confident this is the best approach.

>Given your changes to only allow receive to work on one queue, how do you propose parallel requests will work for a client? One of the reasons we moved to a Redis based system was that we could natively block on multiple parallel responses.

I have discussed this at length with Andrew in Slack. Here's the gist of the conversation:

- Users of service client eventually need to be able to tell the client to "go run these actions in Parallel (using Redis) and return once you have all the responses back." This is different/new functionality from the current standard behavior of "go run these actions in a single job on one server and return once you have the response back."
- The current client handler doesn't support this behavior, but it's intended to be added to it.
- The new Redis transport, as written, supports this. There is no need to receive on multiple queues in order to support this behavior. The request ID encoded into messages (`{'request_id': request_id, 'meta': meta, 'body': body}`) is all that is needed to support this behavior (I encourage @bgreenberg-eb to speak up if he disagrees). It would work as follows:
  - The client, instead of sending one message to a server queue with one job and multiple actions, would send multiple messages/jobs to the server queue, each with a different request ID. 
  - Multiple servers would pop these requests from their queues and handle them simultaneously, then put the responses back on the single client queue.
  - The client would read all the responses back from the client queue until it had all the responses for the given set of request IDs it sent and would return them in the order of request ID.

**Comment by bgreenberg-eb** [comment]
Parallel requests require only that the request ID be returned with each response, so callers can pair responses to requests. This implementation fulfills that requirement.

**Comment by bgreenberg-eb** [comment]
I share @andrewgodwin's mild discomfort with mocking Redis, but if you're confident that it works, there's no problem with it in principle. Since I won't be maintaining this anymore, I defer to your judgment here.

**Pull request merged**